### ISSUE TITLE
Error: Debug Failure. False expression. in `computePositionOfLineAndCharacter` during `getApplicableRefactors`

### ISSUE THREAD

**Issue by aozgaa** [Microsoft/TypeScript#20798]
tsserver version: 2.6.1
hitting sessions: 2512
stack:
```
Error: Debug Failure. False expression.
    at Object.computePositionOfLineAndCharacter (tsserver.js:10652:22)
    at TextStorage.lineOffsetToPosition (tsserver.js:81564:31)
    at ScriptInfo.lineOffsetToPosition (tsserver.js:81784:41)
    at IOSession.Session.getStartAndEndPosition (tsserver.js:86542:48)
    at IOSession.Session.extractPositionAndRange (tsserver.js:86462:35)
    at IOSession.Session.getApplicableRefactors (tsserver.js:86473:31)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (tsserver.js:85436:61)
    at tsserver.js:86669:88
    at IOSession.Session.executeWithRequestId (tsserver.js:86660:28)
    at IOSession.Session.executeCommand (tsserver.js:86669:33)
    at IOSession.Session.onMessage (tsserver.js:86689:35)
    at Interface.<anonymous> (tsserver.js:87881:27)
    at emitOne (events.js:96:13)
    at Interface.emit (events.js:191:7)
    at Interface._onLine (readline.js:241:10)
    at Interface._normalWrite (readline.js:384:12)
    at Socket.ondata (readline.js:101:10)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20)
```

**Comment by mhegazy** [comment]
The `Source: Telemetry` have limited information. we only have the call stack from the crash report. We need to either mitigate the crash, or add additional asserts to get more information in future crash reports.

**Comment by amcasey** [comment]
This looks like it's in the code that maps between offset and line/column.  @andy-ms do you want to take a look?

**Comment by andy-ms** [comment]
Looking at the stack trace, we are getting a call to `getApplicableRefactors` with a range whose `startPosition` is outside of what we believe to be the document text.
Looking at `class TextStorage` it looks like `lineMap` should always be up-to-date with `text`. It's possible that there's some string such that `computeLineAndCharacterOfPosition` on `getLineStarts` with a valid position produces an error, but it seems unlikely. More likely the editor is asking us for refactors at an invalid position.
@aozgaa Is this telemetry from visual studio or vscode?

**Comment by aozgaa** [comment]
All of the Source: Telemetry reports are from vscode. I'll note that explicitly in the future.

**Comment by amcasey** [comment]
@andy-ms @mjbvz I think there was some mention last week that VS Code might ask TS Server about versions previous (or future?) versions of documents, resulting in this kind of problem.  Am I recalling correctly and, if so, do we have a solution in mind?

**Comment by andy-ms** [comment]
That problem was for codefixes, since vscode still asked for code fixes for errors whose locations were now invalid. Something happening during refactors would be a new issue.

**Comment by mhegazy** [comment]
closing for now. we believe this should be fixed. closing for now, and waiting on new telemetry.
